File: advent_of_code/5/advent_5_python/CrateMover.py
from Move import Move


class CrateMover:
  def __init__(self, num_of_stacks: int, moves: [Move] = []):
    self.stacks = self.initialize_stacks(num_of_stacks)
    self.moves = moves

  def initialize_stacks(self, num_of_stacks: int):
    return [[] for _ in range(num_of_stacks)]

  def addCrateToStack(self, line: str):
    j = 0
    i = 0
    while i <= len(line) - 3:
      x = line[i:i + 3].strip()
      substr = x
      if (substr):
        self.stacks[int((i - j) / 3)].insert(0, substr)
      i += 4
      j += 1

  def do_move_9000(self, move: Move):
    for _ in range(move.quantity):
      self.stacks[move.to].append(self.stacks[move.fron].pop())

  def do_move_9001(self, move: Move):
    num_to_move = move.quantity
    aux_stack = [self.stacks[move.fron].pop() for _ in range(num_to_move)]
    self.stacks[move.to].extend(aux_stack[::-1])

  def __str__(self) -> str:
    return str.format("stacks {}, moves {}", self.stacks, self.moves)

File: advent_of_code/5/advent_5_node/CrateMover.js
export default class CrateMover {
  constructor(numOfStacks, moves = []) {
    this.stacks = this.initializeStacks(numOfStacks);
    this.moves = moves;
  }

  initializeStacks(numOfStacks) {
    return Array.from({ length: numOfStacks }, () => []);
  }

  addCrateToStack(line) {
    for (let i = 0, j = 0; i <= line.length - 3; i += 3, j++) {
      const substr = line.substring(i, i + 3).trim();
      if (substr) {
        this.stacks[Math.floor((i - j) / 3)].unshift(substr);
      }
      i++;
    }
  }

  doMove9000(move) {
    const { quantity, from, to } = move;
    for (let i = 0; i < quantity; i++) {
      this.stacks[to].push(this.stacks[from].pop());
    }
  }

  doMove9001(move) {
    const { stacks } = this;
    const { quantity, from, to } = move;

    const movedItems = [];
    for (let i = 0; i < quantity; i++) {
      movedItems.push(stacks[from].pop());
    }

    for (let i = 0; i < quantity; i++) {
      stacks[to].push(movedItems.pop());
    }
  }
}
